:sparkles: Adds kubeadm control plane scale up/down

dlipovetsky opened the pull request:
**What this PR does / why we need it**:
Kubeadm control plane scale up and down, as described in #2241

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2241, #2016 

I'm replacing the scale up unit tests because with the health checks introduced for scale up, they would become integration tests. For the new scale up/down unit tests, I'm going to fake the management cluster.

I'm also replacing the delete unit tests, because the existing ones rely on scaling up the cluster. I'll replace them with a test that works against a static set of machines.

/assign @chuckha @detiber @randomvariable

detiber commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -347,28 +363,60 @@ func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(ctx context.Context,
+	ownedMachines, err := r.managementCluster.GetMachinesForCluster(ctx, clusterKey(cluster), internal.OwnedControlPlaneMachines(kcp.Name))
+	if err != nil {
+		return err
+	}
```
It might be good to check to see if any of the Machines already have a DeletionTimestamp here, and if so return early.

dlipovetsky replied:
Great idea, will add

chuckha commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -251,29 +258,38 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
-			r.recorder.Eventf(kcp, corev1.EventTypeWarning, "FailedInitialization", "Failed to initialize the control plane: %v", err)
+			logger.Error(err, "Failed to initialize control plane")
+			r.recorder.Eventf(kcp, corev1.EventTypeWarning, "FailedInitialization", "Failed to initialize cluster %s/%s control plane: %v", cluster.Namespace, cluster.Name, err)
 			return ctrl.Result{}, err
```
This comment applies to all three cases, initialize, scale up, scale down:

We don't want to retry the health checks immediately. Eventually controller-runtime gives up and does not retry when an error is returned. This means that after 10 failed health checks (which will happen fairly quickly) we have to wait for a full re-reconcile to try again. This is bad because that's a 10 minute pause when trying to scale up.

Because of this, all three cases should instead return `return ctrl.Result{RequeueAfter: 5 * time.Second}, nil` (or whatever duration makes sense, i think even a 10-20 second duration would be appropriate here.

dlipovetsky replied:
Thanks a lot for explaining this.

I think I observed that behavior in my CAPD cluster yesterday. but there was nothing in the log. Would it help to log when the controller is waiting for a full re-reconcile?

chuckha replied:
this wouldn't be possible from the controller, we'd have to change controller-runtime to do this. But generally, the reconciler is always in a state of waiting to do something. A reconcile loop is fairly quick so you can just kind of assume it's always waiting for a reconcile. It's impossible to predect when it's going to reconcile next because in the time it's waiting for a re-reconcile due to the 10 minute time out, some object could change triggering a reconcile anyway.

dlipovetsky replied:
Done.

```
	// HealthCheckFailedRequeueAfter is how long to wait before trying to scale
	// up/down if some target cluster health check has failed
	HealthCheckFailedRequeueAfter = 20 * time.Second
```

dlipovetsky commented on util/util.go:
```diff
@@ -410,3 +410,15 @@ func IsPaused(cluster *clusterv1.Cluster, v metav1.Object) bool {
+func (o MachinesByCreationTimestamp) Less(i, j int) bool {
+	if o[i].CreationTimestamp.Equal(&o[j].CreationTimestamp) {
+		return o[i].Name < o[j].Name
+	}
```
\<pedantic>
Assuming the Name is unique, which it is, as long as we're comparing within one namespace, we need to compare it in order to guarantee total ordering. Without comparing by Name, the sort output will depend on in the order of inputs with equal timestamps, since `Before` returns `false` if the timestamps are equal.
\</pedantic>

(Now I'm wondering if we also need to compare on the Namespace... or the UID (but that isn't available until after the first round trip to the API) :joy:

detiber replied:
It might get a bit stickier than that, since we are using `sort.Sort` not `sort.StableSort`. The results wouldn't necessarily be in the same order as the inputs.

That said, for our use case and since we are going to be enforcing serial scale up, I'm not sure it really matters much. Our use of oldest is mostly just a guess at the one that should be removed.

vincepri replied:
I'd use `sort.Slice` instead of creating yet a new type here

detiber replied:
We'll need this sort in upgrade as well, so I think the type is fine.

vincepri replied:
The function can be reused. This is minor and can be fixed later

dlipovetsky commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -669,3 +717,11 @@ func clusterKey(cluster *clusterv1.Cluster) types.NamespacedName {
+	if len(machines) == 0 {
+		return clusterv1.Machine{}, errors.New("no machines given")
+	}
+	sort.Sort(util.MachinesByCreationTimestamp(machines))
```
Thanks! Credit goes to @k4leung4, who took this approach for MachineSets (https://github.com/kubernetes-sigs/cluster-api/commit/e929a2fc09f20cbf2f1f4e80146257ff69456d0b)

dlipovetsky commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -347,28 +363,60 @@ func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(ctx context.Context,
+	bootstrapSpec := kcp.Spec.KubeadmConfigSpec.DeepCopy()
+	bootstrapSpec.JoinConfiguration = nil
+	err := r.cloneConfigsAndGenerateMachine(ctx, cluster, kcp, bootstrapSpec)
+	if err != nil {
```
Will fix

dlipovetsky commented:
> The changes to go.mod and go.sum should be reverted. We want client-go 0.17.2 and the framework should not be importing CAPD

I'll revert that. What do I do about it locally, though? That's the diff after running `make modules` (go 1.13.8) :confused:
```
$ git status --untracked-files=no
On branch controlplane-scale
nothing to commit (use -u to show untracked files)

$ make modules --debug
GNU Make 4.2.1
Built for x86_64-redhat-linux-gnu
Copyright (C) 1988-2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Reading makefiles...
Updating makefiles....
Updating goal targets....
 File 'modules' does not exist.
Must remake target 'modules'.
go mod tidy
cd hack/tools; go mod tidy
cd test/framework; go mod tidy
go: finding sigs.k8s.io/cluster-api/test/infrastructure/docker latest
cd test/infrastructure/docker; make modules
GNU Make 4.2.1
Built for x86_64-redhat-linux-gnu
Copyright (C) 1988-2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Reading makefiles...
make[1]: Entering directory '/home/dlipovetsky/projects/cluster-api/test/infrastructure/docker'
Updating makefiles....
Updating goal targets....
 File 'modules' does not exist.
Must remake target 'modules'.
go mod tidy
cd hack/tools; go mod tidy
Successfully remade target file 'modules'.
make[1]: Leaving directory '/home/dlipovetsky/projects/cluster-api/test/infrastructure/docker'
Successfully remade target file 'modules'.

$ git status --untracked-files=no
On branch controlplane-scale
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   test/framework/go.mod
        modified:   test/framework/go.sum
        modified:   test/infrastructure/docker/go.mod
        modified:   test/infrastructure/docker/go.sum

no changes added to commit (use "git add" and/or "git commit -a")

$ go version
go version go1.13.8 linux/amd64
```

vincepri commented:
Ping when you're ready for another set of 👀

chuckha commented:
> > The changes to go.mod and go.sum should be reverted. We want client-go 0.17.2 and the framework should not be importing CAPD
> 
> I'll revert that. What do I do about it locally, though? That's the diff after running `make modules` (go 1.13.8) 😕

That's very weird. On my system, that is not the case. Once I revert those changes and run make modules, go.sum cleans up and my changes persist 🤔 

Let's see if prow complains after you revert them. It will complain about outdated files if go modules is not up to date.

dlipovetsky commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -669,3 +717,11 @@ func clusterKey(cluster *clusterv1.Cluster) types.NamespacedName {
+
+func oldestMachine(machines []clusterv1.Machine) (clusterv1.Machine, error) {
+	if len(machines) == 0 {
+		return clusterv1.Machine{}, errors.New("no machines given")
```
https://github.com/kubernetes-sigs/cluster-api/pull/2336

dlipovetsky commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller_test.go:
```diff
@@ -1382,184 +1196,193 @@ func TestReconcileControlPlaneDelete(t *testing.T) {
+
+		controlPlaneMachines := clusterv1.MachineList{}
+		g.Expect(fakeClient.List(context.Background(), &controlPlaneMachines)).To(Succeed())
+		g.Expect(controlPlaneMachines.Items).To(BeEmpty())
```
Good catch, will fix, thanks

chuckha commented:
/approve

/assign @vincepri @detiber 
for lgtm

vincepri commented:
Reviewing

vincepri commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -346,28 +363,67 @@ func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(ctx context.Context,
+		return ctrl.Result{}, errors.Wrapf(err, "failed to create control plane Machine for cluster %s/%s", cluster.Name, cluster.Namespace)
+	}
+
+	return ctrl.Result{Requeue: true}, nil
```
Add a comment explaining why the need for `Requeue: true`

dlipovetsky replied:
Will add comments

vincepri commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -346,28 +363,67 @@ func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(ctx context.Context,
 	}
 
-	return kerrors.NewAggregate(errs)
+	return ctrl.Result{Requeue: true}, nil
```
Add a comment explaining why the need for `Requeue: true`

vincepri commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -346,28 +363,67 @@ func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(ctx context.Context,
+		return ctrl.Result{}, errors.Wrapf(err, "failed to delete control plane Machine %s/%s", machineToDelete.Namespace, machineToDelete.Name)
+	}
+
+	return ctrl.Result{Requeue: true}, nil
```
Add a comment explaining why the need for Requeue: true
